### PR TITLE
Add Typescript typings + rework

### PR DESCRIPTION
**Pull request by kanongil**
I wanted to add official typings to this module. In order to best do this, I **added 2 breaking changes**.

1. Move default export to `Mimos`. Ie. the class will now be found at `Mimos.Mimos()`. This is because Typescript has issues with default exports, and matches what was done in `Boom`.
2. Change `path()` to return `null` when no entry is found (instead of `{}`) – The `{}` return was a nonsensical API. Eg. declaring the return value with type `MimosEntry | {}`, makes it impossible to narrow it. With `null` it can be done with a simple `if (returnedValue)`.

While I was doing the changes, I found opportunities to speed it up. From the DB setup, the initial load took 10s of ms on my machine. I made several changes to speed it up:

1. Use `Map()` type instead of POJO for DB storage.
1. Rework compilation using knowledge of source data to avoid expensive `clone()` and `merge()` of mime-db data.
1. Use initially compiled DB as basis for overrides (instead of a full re-compile).

The behavior should match the current for common usage, with 2 exceptions for corner cases:

1. The `extensions` property is assigned with the value from the parsed json. If someone were to modify this mime-db object later, it can change the returned `extensions` value. *I don't believe this will be an issue in practice.*
1. The `path()` lookup might match another (valid) mime-type, if an override was provided for it, and multiple entries share an extension. *This is a rare case, that already relies on V8-specific behaviour. It probably works better, since overrides now **always** take precedence.*

For the ´type()` lookup, I made 2 speed-ups:

1. Faster type extraction from provided mime type (removing any options), using `indexOf()` and `slice()`.
1. Attempt a pre-check with a quick lookup, before doing the expensive `.trim().toLowerCase()`. This is a heuristic based optimisation, which will make the lookup slightly slower for an upper-cased type lookup, but make it a lot faster for the simple common case.

Since `mimos.type()` is called on most hapi server responses, this should make hapi slightly faster.

### CODE REVIEW

**Comment by devinivy** [comment]
@kanongil thanks for this.  I will review it more closely soon, but this all generally sounds good to me.  While I agree with the breaking change regarding the result of `path()`, I wonder if we should temporarily hold off on this since it is part of hapi's public API via `server.mime`.

**Comment by kanongil** [comment]
Hmm, I had not thought about the direct exposure in hapi. Technically, it would not be considered a breaking change from hapi POV, as it is just defined as whatever object mimos created, but I would prefer not to push it.

I guess I will go back to the `MimosEntry | {}` for now, and TS users will just have to live with it being a pain to use?

It would be handled something like this, for somewhat safe typing:

```ts
const entry = mimos.path(path);
if ((entry as MimosEntry).type) {
    const mime = entry as MimosEntry;
    …
}
```

File: lib/index.js [hunk 1]
```diff
@@ -6,87 +6,142 @@ const Hoek = require('@hapi/hoek');
...
-        if (mime.predicate) {
-            return mime.predicate(Hoek.clone(mime));
+exports.Mimos = class Mimos {
+
+    #db = internals.base;
+
+    constructor(options = {}) {
+
+        if (options.override) {
+            Hoek.assert(typeof options.override === 'object', 'overrides option must be an object');
```
Comment (Nargonath) [inline lib/index.js:85]: Shouldn't we use `@hapi/validate` instead of `Hoek.assert`?

**Review by Nargonath — COMMENTED** [review]
LGTM but it might be a good idea to release progressively with a non-breaking version first. Also shouldn't we create an issue on `hapi` or create a new major milestone to keep track of all the breaking changes we want to do in the next big version? I believe this is not the first breaking change we postpone and I wonder if we're not going to forget about them as time goes by.

File: lib/index.js (hunk 1 above)
Comment (kanongil) [reply lib/index.js:85]: `@hapi/validate` is generally not used to parse simple options due to the significant processing & memory overhead. This means it is almost never used for potentially hot code paths. For once-only initializers like this it is a judgement call.

In this case I think it is better as-is. This also avoids loading an expensive dependency when used outside the hapi eco-system.

Comment (Nargonath) [reply lib/index.js:85]: Good point @kanongil. Indeed it feels as an overkill now that I think about it.

**Comment by Marsup** [comment]
Wouldn't we benefit from having a kind of citgm for hapi that we could trigger with a matrix of modules and see if at least it builds?

**Comment by devinivy** [comment]
@Marsup yes, that would be amazing.

**Comment by kanongil** [comment]
I have given this some further thought, and if we want to go with this change, I would be inclined to just update hapi with it _without doing a breaking change_. As I already said, this is technically not a breaking change for hapi, but more importantly, even if pushed as a breaking change, it can still cause issues for years since plugins might call into the api, expecting the old behavior. The returned value is a corner-case that could go easily pass 100% test coverage even after such an update.

So I guess the question is, if this instability is worth it. I'm leaning towards: NO. I would rather keep the existing API, and potentially take measures to simplify TS usage of the returned object.

File: test/index.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+expect.error(mimos.path(new URL('file:///my/file.html')));
+
+// MimosEntry
+
+const entry = mimos.path('/my/file.html');
+if (entry instanceof MimosEntry) {
+    expect.type<MimosEntry>(entry);
+}
+
+//expect.error(new MimosEntry());       // Test is disabled due to Lab issue
```
Comment (kanongil) [inline test/index.ts:47]: If uncommented, lab incorrectly errors:

```
Types:
	test/index.ts:42:13: Constructor of class 'MimosEntry' is private and only accessible within the class declaration.
	test/index.ts:42:0: Expected an error
```

Ie. it reports a type error, and then reports an error because there was no type errors…

**Review by kanongil — COMMENTED** [review]
I undid the `path()` breaking change.

To make TS usage simple, I export the `MimosEntry` class and expose it as such in the types. This enables a simple `instanceof` check on the the returned value, which will correctly narrow the type. Eg.:

```ts
const mime = mimos.path(path);
if (mime instanceof MimosEntry) {
    /* mime is ready for use */
    …
}
```

File: test/index.ts (hunk 2 above)
Comment (devinivy) [reply test/index.ts:47]: I took at look at this, and believe we have a fix: https://github.com/hapijs/lab/pull/1007

Comment (devinivy) [reply test/index.ts:47]:
```suggestion
expect.error(new MimosEntry());
```

File: package.json [hunk 3]
```diff
@@ -18,10 +19,12 @@
   },
   "devDependencies": {
     "@hapi/code": "8.x.x",
-    "@hapi/lab": "24.x.x"
+    "@hapi/lab": "24.x.x",
```
Comment (devinivy) [inline package.json:22]: This minimum version is necessary because it contains the fix in https://github.com/hapijs/lab/pull/1007 for functionality used by the test suite here.
```suggestion
    "@hapi/lab": "^24.2.1",
```

**Review by devinivy — APPROVED** [review]
Looks good, thanks Gil!  I've tested these changes locally against hapi (with some very minor changes within hapi), and all seems to be well there too.  I also peeled off the breaking change to `mimos.type()` proposed here into https://github.com/hapijs/mimos/issues/38.

File: API.md [hunk 4]
```diff
@@ -3,25 +3,25 @@ Mimos is a convenience class for retrieving mime information objects.
...
         - `key` - the key is the lower-cased correct mime-type. (Ex. "application/javascript").
-        - `value` - the value should an object following the specifications outlined [here](https://github.com/jshttp/mime-db#data-structure). Additional values include:
+        - `value` - the value should be an object following the specifications outlined [here](https://github.com/jshttp/mime-db#data-structure). Additional values include:
           - `type` - specify the `type` value of result objects, defaults to `key`. See the example below for more clarification.
           - `predicate` - method with signature `function(mime)` when this mime type is found in the database, this function will run. This allows you make customizations to `mime` based on developer criteria.
 
 ### `mimos.path(path)`
 
-Returns mime object where:
+Lookup file extension from path and return mime object, or `null` if not found, where:
```
Comment (Nargonath) [inline API.md:19]:
```suggestion
Lookup file extension from path and return mime object, or an empty object literal `{}` if not found, where:
```

I believe we forgot to update the documentation after the breaking change revert.

**Review by Nargonath — COMMENTED** [review]
Found just a little error in the docs otherwise LGTM

**Review by Nargonath — APPROVED** [review]
